### ISSUE TITLE
Armory 2020

### ISSUE THREAD

**Issue by luboslenco** [armory3d/armory#1545]
Welcome! With 2020 already in full swing, I would like to outline some plans for Armory architecture in the upcoming year. 🔨 

- Move rendering to [Graphics5](https://github.com/Kode/Kinc/tree/master/Sources/kinc/graphics5). G5 is a newer API being developed in [Kinc](https://github.com/Kode/Kinc), leveraging modern graphics APIs like D3D12 / Vulkan / ..
- As a result, support for D3D12, Vulkan, Metal and WebGPU will be priority. Once running smoothly, older graphics APIs will be dropped.
- Implement ray-tracing for dynamic scenes. Right now Armory already has [DXR support](https://github.com/Kode/Kinc/blob/master/Backends/Graphics5/Direct3D12/Sources/Kore/RayTraceImpl.cpp), but only handles static scenes. The goal is to have [tanks demo](https://github.com/armory3d/armory/wiki/tanks) running on a ray-tracing render path.
- Move low-level parts of Armory (like [iron](https://github.com/armory3d/iron)) to C. This is to better take advantage of multi-threading coupled with Graphics5 API. As a result, Armory will not be dependant on Kha (Haxe), but will interface directly with Kinc (C). An [extended version of Krom](https://github.com/armory3d/Kromx) which exposes Armory-specific functionality from C to Haxe/JS will be developed to accomplish this.
- Armory traits will be written in Haxe/JS like usual, or anything which compiles into WebAssembly.
- For web deployment, WebAssembly and WebGPU will be used.

Feel welcome to bring up any of these points for a discussion. If you have additional ideas which may improve the project further, please bring those up as well. The goal is to keep Armory viable long-term with a modern base ready for upcoming years.

My biggest thanks to everyone who already contributed to the project in any form over the years!

**Comment by Clockwork-Muse** [comment]
... if you're moving things to C, would Rust be considered?

**Comment by nanjizal** [comment]
I can't seem to get any online samples of WebGPU working on my mac :(

NME uses Webassembly http://hughsando.com/category/nme/ it does not load or run as fast as WebGL ( see my PolyominoTriangles tetris tests from 2017, in Kha WebGL, WebGL haxe, WebGL Heaps, WebGL OpenFL - NME JSprime equivalent ). 

WebGL is definitely the ideal output for showing potential clients work as it runs on mobile from a github repo, and I am yet to believe the Webassembly / c in browser offers enough advantages.

I am sad that Armory is moving away from Haxe. If Iron moves to C, it will definitely be beyond many of us experimenting with, for instance projects like Rice2D use your components and could use Iron.

I looked at Rust briefly once meh, Nim seems more interesting and looks like there may already be start of a Nim target for Haxe, certainly I would suggest using Nim over c if you want more reliable code base, it has good parallel support and speed.
https://github.com/RapidFingers/Craxe

Haxe has many other evolving target solutions cppia, Hashlink, LLVM target etc...and they involve many talented devs, I think it's nice that Armory currently has potential Haxe choices but in future you will be more tied to c tech. For instance I presume if Robert stopped working on Krom/Kinc/Kinx then Kha could be maintained and run on another target - same type of argument that users had against Haxe till other devs got involved.

The other consideration you need more stability, porting Iron to c is only going to divert you from improving current reliability and stability, also it may absorb a lot of your time on complex stuff and not give you as much time for the visual graphics stuff you also love?

But these are just my ideas and fully aware that I am very biased being a haxe user, and I don't get much time for code let alone learning blender so my opinion is not that relevant, but always amazed by what you create.

**Comment by BlackGoku36** [comment]
@Clockwork-Muse Kha is to Haxe, Kinc is to C but there is nothing to Rust, If kinc wrapper was made for rust, than you would have to write entire thing in unsafe blocks, which defeats purpose of rust(AFAIK).

---

```
I am sad that Armory is moving away from Haxe. If Iron moves to C, it will definitely be beyond many of us experimenting with, for instance projects like Rice2D use your components and could use Iron.
```
🤔 

---

What about projects like zui, It will still stay in Haxe, right?

---

With iron and armory written in C, it would open up lot of possibility, like use C's projects like Imgui, etc.

**Comment by trsh** [comment]
Can't hide that I am excited. I do support all the directions you are aiming, but have some suggestions:

a) I don't think all of the bugs registered are because of the current low-level parts tech stack and will magically solve as you move to Kinc. Probably lot's are logic bugs and things that just need rework. So would be great if along the rewriting you pick the most urgent ones (like lightning) and solve them, so we don't get the new version with same problems. Maybe some of them can be solved even before the big migration, so we can finish our games and demos?

b) To attract more contributors please implement some "best practices" for coding. Managing the code in a manner as efficient as possible, to make it easily understandable, by the person that writes the code, and by the people that will read the code in the future.  There are many books and online material on the topic, but what can I say from my exp:

1) Always go for hard OOP approach and split the logic as needed in sub sections, so there are less huge code chunks, that are hard to extend.
2) Always comment the code appropriate 
3) Check some best practices for namings. Like use camelCase for this, pascalCase for that, always start a boolean variable name with `is`, always write understandable function/variable names, etc etc. Maybe there is even some automated linting for C, like for JavaScript. I don't know.
4) Don't ignore warning

Reading clean code, is like reading a good book and I think is very important for open source projects.

c) Update your blog or twitter even with small accomplishments , so we can fallow the development and maybe help.

d) Please, don't aim to much for high end gear and tech, like RTX graphic cards for ray-tracing (ctytek and Godot for example implemented ray-tracing on Vulkan, with no need of DXR - RTX), VR's, etc. But focus more on stabilizing the engine basics. Most devs here are not Big companies that want to make the next FPS AAA title, but rather small Indy developers why are aiming for simple Mobile/web game. Everybody tells me that Armory already has fantastic looks. What people are mostly missing is like Network stuff, online streaming for large worlds, adding post processing or custom shaders with ease, ik resolving for procedural animations, more of the blender shader nodes support (like refraction), etc., But MOSTLY the engine overall stability for basics. If you don't believe me, create a survey or just check the feature requests in git :). And we still want graphics scalability so we can Im games also for older devices.

I hope this helps, good luck!

**Comment by RobDangerous** [comment]
Cool Spot, looking forward to Kinc-Armory!

**Comment by Sandy10000** [comment]
I am very relieved to hear the future of Armory. 

Will bugs related to kha and iron be fixed in the future? Or are you dedicated to building a new environment?

**Comment by Clockwork-Muse** [comment]
@BlackGoku36 - Yes, calling directly into it would require `unsafe` blocks.  However, dependent code would not - depending on the exact structure this could be a significant portion.

**Comment by Didiercode** [comment]
It’s normal that the goal is to keep Armory viable long-term with a modern base ready for upcoming years… but as Armory is here since several years without switching to version 1, I can only ask you the question if, quite frankly, isn't this actually the classic target drift that will again unnecessarily delays first reliable version of Armory, drives up costs/time which you may not have enough to do and can even destroy your project. As you are the only developer/designer, it’s essential to avoid this risk. 

Otherwise from my personal point of view, I would be very happy to use D3D12/DirectML in Armory.

**Comment by RobDangerous** [comment]
Don't spoil the fun... because for a one person project fun is the key to getting work done efficiently. Being on top of the game in terms of graphical finesse seems to be what Lubos enjoys most about this project from what I can tell - please let him just do that and I'm condident that'll actually minimize the risks.

**Comment by Didiercode** [comment]
the fun of a project is the key to effective work ... up to a certain point; it's called the age of heroes and effectively it's more fun ... but it works for a while. The investors/supporters/ecosystems...  find the fun too in software finishers who don't have as much fun.

**Comment by RobDangerous** [comment]
Urks. Anyways, Lubos, I have more advice for you: Get away from that traditional versioning scheme. Nobody ever asks me for version 1.0 because I'm already at version 2020.1. It makes my life better.

**Comment by trsh** [comment]
I must agree with @Didiercode . Every project is fun to certain point. After that is hard work and dedication to bring something to the finish line. New ideas, tech, etc. often kill projects or keep them in never ending loop. If you are building a engine for your self, its fine, but there we have supporters and donators and whole community behind. We want to make actual games with Armory, not just experiment with shiny features :)

**Comment by VicentGJ** [comment]
I agrede with @RobDangerous in this. When it comes to this kind of projects the important is that the maintainer found it fun to work on it, or they are driven by a very particular necessity. If the project proof to be useful to other people, and start growing, then the motivation will switch from fun to the happiness that comes from knowing  that your work is beneficial to a lot of people. That is the way this usually works and because of that we have to support @luboslenco in any technological decision he made.

**Comment by BlackGoku36** [comment]
Writing a game engine isn't small thing, neither is implementing stuffs like RTX, they take time, sweat, tears(bugs lol), blood(ofc not), and headaches to be implemented. The ability to achieve this is largely governed by how much fun you get working on this, believe me you can't work on stuffs you don't find fun. In our case, we know that @luboslenco find stuffs such as fancy looking shaders, modern technologies, etc. fun. So, I agree with @RobDangerous and @VicentGJ .

To be honest, armory already looks like "1.0 stable" game engine, with which you can make complete game in. The only thing right now, is that it is full of bugs, and it feel like whole engine is holding up with duct tape. So, addressing bugs should be top-priority right now, ofc i know that bugs are really headache to solve, so since we are are porting Armory from Haxe to C (Kinc), I think we should........re-write the whole engine from ground up? This will solve many bugs such as lightings, animations, and also introduce better code base architecture, and the main goal should be to restore all the promised stuffs that were removed because of reasons, like blender viewport rendering, Sub-surface scattering, Node based render-path, grease-pencil(we already got @N8n5h taking care of this), etc. With C, we will able to use many C libraries and remove codes that are re-write in Haxe, this will help a lot with fixing bugs and adding of new features.

**Comment by trsh** [comment]
If I put all opinions together, I recommend the fallowing:

1) Move to Kinc (C) - rewrite clean robust code with "best practices", but don't reinvent the wheel where it's not needed, so it's not like 2 years orsmth. 

2) Address prioritized bugs and community feature requests/promises **over** fancy RTX, VR, etc stuff in respect to all donations and support. I even suggest to open a Pool of the most anticipated bug fixes and features for the upcoming release.

@BlackGoku36 

> To be honest, armory already looks like "1.0 stable" game engine, with which you can make complete game in. 

Depends on what games. Remember our drone demo hanging for a Year. Go and finish it! :P

> ...I suggest we........re-write the whole engine from ground up? This will solve many bugs such as lightings, animations, and also introduce better code base architecture...

I don't think every piece will be reinvented from ground up (to consuming), I think most of logic will be **Ported**. That, if it isn't Kha issue, will leave the same problems pending.

Anyway, time to learn C

**Comment by Didiercode** [comment]
@trsh sure that a real disruptive approach sould be by redisigning  Armory completely around AI and path tracing and seems to me too the best way for getting more fun !  

Comparatively with the approach proposed by @lubos with Kinc / C, it seems that we are stuck to ideas and concept based on evolutions of existing data representations, coming from an outdated trick based on 80's approaches/ technological foundation from work on CAD software which were never meant nor designed to be used at  using today IA/ML technics. 
Thus this needs building a new type of data representation that Kinc and C are not prepared/made for that I suppose too. Compared to a boring rewriting Armory code, that's really what could be fun to do. 
@trsh your link to the http://www.lightmass-dynamics.com/  is a prefect illustration of what could be a real fun way to get Armory viable long-term with a modern base ready for upcoming years…

**Comment by BlackGoku36** [comment]
> Depends on what games. Remember our drone demo hanging for a Year. Go and finish it! :P

Care to point out what was wrong? Well, I can. The bug we faced were:
* Animations (we could not add those tentacles things because animations issue with bones and stuffs, so we had workaround our way with programmed arm moment)
* Emission issue
* Lightings issue (Area lights would not work)
* Environment map issue (Texture looked displaced and didn't contribute to light correctly)
* Rotations (Had sudden spike moment when starting mouse moments, observable in official templates)
Tell me thing that made it look like it was not "1.0"

Careful people, we got some 10x programmer who have coded game engine and have made really good contribution to Armory itself

**Comment by trsh** [comment]
> Tell me thing that made it look like it was not "1.0"

Yeah, not working lights doesn't sound like 1.0 version :P to me (but that's opinions). Try to play with Omni, add multiple, also doesn't work :P. And you didn't mention all issues. Anyways, because of those NOT fancy things, but basics we could not finish the demo, so that's not 1.0 to me, and I stick to it.

**Comment by trsh** [comment]
@Didiercode tere is no more info of the tech as that site and videos. And it's very experimental. Way to far to aim for :P

**Comment by Didiercode** [comment]
If Armory was managed like a startup and taking the way to become disruptive with IA inside, it could be supported like Lightmass Dynamics with the Techstars in Canada or France  ... look info here https://www.techstars.com/content/accelerators/dix-nouvelles-startups-accelerent-leur-developpement-avec-techstars-montreal-ai-2019/ 

Thus Armory could get more fun and with ressources with Techstars as it is an acceleration program based on mentoring. Many of these mentors are entrepreneurs themselves, but a large number of IA technical experts and business people from a variety of industries.

**Comment by BlackGoku36** [comment]
> I think that @luboslenco can realize now in retrospect that wanting to rewrite alone the whole Armory game engine as he proposes today, he would end up with decades old technology.

Do tell me how?

**Comment by Didiercode** [comment]
@BlackGoku36  look at retrospect of releases first ![image](https://user-images.githubusercontent.com/20455592/72436478-c9e8e580-37a0-11ea-906c-75d4f785fb04.png)
 and tell me how could be the future planning ... everymonth delivery at start then every year ... first version in 2030 with KinC ?

**Comment by BlackGoku36** [comment]
I don't get what you are trying to say. You are talking about lubos using decades old technology and now you are talking about Armory's with kinc releasing in 2030

**Comment by trsh** [comment]
I wonder how we could speed up the development. Can't wait to get my hands on the next version :P

**Comment by Didiercode** [comment]
@BlackGoku36 what we can predict without necesserely being a guru in the domain, is that as game industry is providing VA in industries like training, data sciences, communication, defense, education, medecine, ... this disruptive potential of game engines techs will boost the growth of new types of companies and threaten  well established positions in next years like this UE4 one. 

It's up to Lubos to offer a real value proposition that cannot be dismissed in the next decade... but I don't see it today and I hope to be wrong as he's already achieved something surprising with Armory 3D and ArmorPaint for someone on his own.

**Comment by arigbs** [comment]
> Armory traits will be written in Haxe/JS like usual, or anything which compiles into WebAssembly.

Blazor?

**Comment by smxhams** [comment]
Love your work Lubos.

**Comment by onelsonic** [comment]
New Year, New stuff, Always great news to hear new stuff! @lubos if you need any help, shout-out! 
Not sure if there are a lot of coders amount us?
C language equal SPEED SPEED and SPEED, and... access to all the major libs both free and commercial.
So we will get up to date physic engines to plug-in and much more.
Not sure about the multi-threading as that will be dependant upon the target?

I'm more interested in knowing the next planned workflow.
Will it be similar to current:
Coding in Haxe/C > that is then transcoded into the target ie WebGPU, Metal ?

Then speaking of Armory, correct me if I am wrong as I started to asses the tech 6 months ago, since August so don't know the full history.
The current and planned workflow are both heavily depend on Kha/Kinc.
Which is currently targeting a lot of systems but have some core critical issues remaining.
It feel like trying to target to much systems none are fully polished and finished. Node JS/Krom are probably at a better state than other targets but still incomplete.

It seems like Kha/Kinc is the bottleneck here, the problem is that both projects have different use it seems. And one need the other to function, as Armory is based on Kha/Kinc.

For example, when reviewing the targets, now with vulkan why still bother targeting Direct3D12 in the future? 
When vulkan is achieving higher FPS vs DX12 on AAA games.

The strength of Armory was
BLENDER
LIGHTWEIGHT
SPEED
MOBILE/PC
CODE READABILITY (HAXE)
which makes it a pretty good tool if a real tech game demo could made.
(Thinkink there of a Godot robot Shooter Demo like)

Then you have the PC vs Mobile paradigm that can define you targeted systems.

As of today Ray Tracing, RTX in Game is more of a marketing gimmick use by one Brand to sell more hardware, like Cuda was.
Still far to be mainstream.
Then the best renderers in this area are PIXAR and OTOY (OctaneRender) but can't be use for a realtime AAA game yet with our current gaming hardware, except if your are willing to spend +50,000$ for your graphic cards.
So one can question how some reverse path tracing technique with a MonteCarlo estimation will perform on mobile?

Seems like this can hurt the strength of Armory capable of targeting Mobiles.

Then moving to C and knowing its complexity you will also hurt the code readability, where HAXE is pretty good at this. OCaml is even better, probably the language requiring the less number of lines for a same program compare the others (this last one is heavenly used in real time finance, but can be applied to gaming)

Being in the tech/gaming industry for +15 years my advise will be to try emancipate from any bottleneck (Forking your own Kha/Kinc) and focus on your current product strength which brought you some visibility and a small community and improve these strength continually while launching a fully finish tech demo and then game.
Otherwise you can end up with a never-ending research on trying to achieve something but never complete it.

If your goal is not a finish product but more the research aspect, then these guys pointing some exotic deep learning renderer is an area with good futur potential.
New data representation or even
Voxels are also and area of research
with commercial product launched recently.
https://www.atomontage.com/
https://www.euclideon.com/

If the focus is gaming you still have the limitations of the current hardware still currently optimised toward raster. John Carmack has made a lot of talks about this.

The essential is knowing what motivate you and keep going at it full speed, with a smile.

Happy New year to you all.
(From the Green Hill Zone)

**Comment by trsh** [comment]
https://www.euclideon.com/ is rather a bad example for gaming industry (they backed of long time ago).  https://www.voxelfarm.com/index.html is a good one, and it can be integrated into Armory and its already used in games. https://www.atomontage.com/ has been into development for eternity, so Im suspicious about they end product for gaming - could end up like euclideon. This another voxel project to fallow https://twitter.com/ProgrammerLin . But yeah, voxels are like different genre with different qualities.

**Comment by RobDangerous** [comment]
Because D3D12 is required for Xbox and consoles are currently my main focus. How would forking Kinc aka effectively halfing the man power working on the backends improve things?

**Comment by onelsonic** [comment]
My comments are for the Armory engine.
My advise toward Armory was to focus on its own targets where it currently has pretty good strengths vs the competition.

The consoles markets already have more than established engines and it will be pretty hard to convince a team to switch to Armory as of today so for Armory is D3D12 required?

Talking about Kha/Kinc and its future development will be a totally different discussion.
I only reviewed Armory current state and capabilities.

Or is Armory development is to support Kha/Kinc and the plan is to launch an alternative to UE4/Unity/CryEngine for the consoles market?
That could be interesting...

**Comment by RobDangerous** [comment]
Armory does not have its own system targets. When you talk about D3D12 and Vulkan, you talk about Kinc. When Armory manages its own fork of Kinc and deletes the D3D12 code that doesn't save the Armory team any work at all.

**Comment by onelsonic** [comment]
Having a priority in 2020 to support D3D12 sure can.

**Comment by RobDangerous** [comment]
But how would forking Kinc help with that?

**Comment by N8n5h** [comment]
I think the fork is for Krom being Kromx and not Kinc exactly, @onelsonic.

**Comment by onelsonic** [comment]
Current Armory version has a lot of issues related to Kha.
Next iteration will use Kinc, similar issues will pop-up eventually and be hanging similar to the multiple lighting issues mentioned above.
Forking a version of Kinc for Armory while focusing only on target systems where Armory has an advantage and benefit to exist can speed up having an operational game engine.

**Comment by RobDangerous** [comment]
Why would you need to fork it to focus on the targets you want to focus on?

**Comment by onelsonic** [comment]
Well, removing the extra steps to validate the compatibility with other unnecessary targets?

**Comment by RobDangerous** [comment]
There are no extra steps.

**Comment by trsh** [comment]
I think we can separate Armory code contribution in 3 sectors (ordered by complexity):

1) New nodes, Traits
2) Changing/fixing the armsdk
3) Changing/fixing Kinc

Me (and some dudes I know) could contribute in sector 2, but there is that problem, that I have to work with Kinc and I quote @RobDangerous : "Kinc is not for beginners. If you can't fix bugs in it, you should use something else.". The standard route would be, if encounter a problem with Kinc, I report a bug and kindly ask Kha/Kinc developers to look at it, but that doesn't seem to work. Instead i'm being humiliated as bad programmer, because I'm not fixing it myself. Another quote of the same discussion: "You don't become a good programmer by asking for bug fixes, you become one by making them - otherwise you're just consumer, not a creator". Button line - if I'm a good Kinc consumer, but no creator, I can't help developing armsdk. And this again leaves @luboslenco own alone working on everything. Kind of sucks :P


"

**Comment by BlackGoku36** [comment]
Oh boy, here we go again.
Time to grab some popcorns 🍿

**Comment by trsh** [comment]
@BlackGoku36 im not into fighting. Just confused and sad

**Comment by luboslenco** [comment]
(@trsh please keep these masquerades away, it's been going on for years now. Basically unable to have a discussion without those popcorn sessions.)

**Comment by luboslenco** [comment]
Thanks to all for posting the feedback, it helps a lot with planning things around.

- For coherence with Kinc we should stick with C instead of Rust. Kinc has a very clean code-base and following that will benefit Armory - because the structure is already laid out. Once Kinc will get Rust bindings we can follow similar steps.
- I think the amount of work / changes proposed in the original post got exaggerated a bit. The hardest part is already done by Robert, Armory already runs on Kinc/C today. What will change is iron gradually moving to C to take advantage of Graphics5 API - and iron is tiny - 10k lines. 🙂 (Even [Kromx](https://github.com/armory3d/Kromx/blob/master/Sources/main.cpp) already runs directly on Kinc/C, instead of Kore/C++)
- Zui and stuff running on Graphics2 will stay as is using Haxe. (For people familiar with Kha/Kinc structure, there will be `Graphics2onGraphics5`!)
- Contributing to the [Kode](https://github.com/Kode) projects is of the highest importance long-term, forking would hamper that. Been working with Robert for the past 6 years and Armory would not exist otherwise. Lots of cool stuff coming from Robert as well which will have impact on Armory, do not want to miss that. 😃 
- Ray-tracing will not be mandatory, it will be one of the available rendering paths. Our ray-tracing implementation in Kinc is not tied to DXR. It just happens that we only have DXR backend as of now (because it is the simplest and most performant one).
- Armory will be kept on [monthly releases](https://github.com/armory3d/armsdk/releases) as it has been for the past few months.
- Will make sure to benchmark and post the results once we get there but I would not bet on `webassembly + webgpu` under-performing our current `js + webgl` solution. webgpu implementation status can be monitored at https://github.com/gpuweb/gpuweb/wiki/Implementation-Status.

Will post another update detailing the progress along the way!

**Comment by trsh** [comment]
@luboslenco there are no masquerades. I just quoted what I got, when I wanted to learn to use Kinc and asked for Help in IRC.. and it's related to this and the armsdk contributing part. We want to get involved and help. Anyways, good luck with the work!

**Comment by RobDangerous** [comment]
(you're all free to read the whole IRC chatlog starting with last month to look behind the masquerade)

**Comment by QuantumCoderQC** [comment]
What would be the state of Physics Engine? Will it still continue to be applied as a Haxe Trait?

**Comment by onelsonic** [comment]
Nice one 👍
if the iron C code is clean already with only +10k lines it should be fine. (just need to limit the inter-dependencies of files and linked imported libraries if any).
As I said I did not know the full history of you working with Robert for the past 6 years, that was just an analysis with an external eye.

With Ray-tracing as a separate Renderer path, your mobile strength should still be there. ✨

My last thoughts on this is your networking layer and how all will interact with the web
(game server / online interactions).

For the NodeJS/HTML5 target JS is pretty easy to interface with the web together with WebGL.
And there are a tone of JS libs that can be use for this purpose.

How about Webassembly? Let's say that someone want to create an HTML5 Online web game with networking.
All the networking layer will have to be written for Webassembly interfacing with WebGPU?
Seems like some hard work ahead?

The work in progress done by Apple on WebKit and Google on Chromium concerning WebGPU is to use their respective Javascript API.
so more > JS + Webgpu.

**Comment by Disar** [comment]
@luboslenco So to be clear only mobile devices with vulkan support are going to work?

**Comment by artanokhin** [comment]
Hello!

Armory 3D is a very large and necessary project. We must try again to return the engine back to the core of the program. Improving the game engine, we improve the blender - improving the blender, improve the engine. There is no need for a new version to start by default from the list: "General, 2D Animation, Sculpting, VFX, Video Editing and **Game developing** - with correct setting windows for work!

But the most important thing: Please pay attention to the ease of use of the engine.It should be included in the Blender kit or it may be outside of its environment, but it may just be well compatible, which allows you to quickly export materials and settings from Blender to the Armory 3D engine, similar to Armory Paint. But visually it is necessary to streamline everything. For example, so that all mechanics can be implemented through a node system or through writing code. And for each of the options has its own environment, its own interface.

There needs to be a significant shift in how the end user uses the engine. Similar as it was with Blender 2.4 and 2.5

I think that close collaboration with Blender developers is needed.

P.S. My craziest idea. This is the ability to create scenes and game mechanics in Blender and then transfer to Unreal Engine or Unity or Godot with all the settings and working mechanics.

**Comment by SonKim** [comment]
> d) Please, don't aim to much for high end gear and tech, like RTX graphic cards for ray-tracing (ctytek and Godot for example implemented ray-tracing on Vulkan, with no need of DXR - RTX),

Crytek and Godot ray tracing implement that don't support hardware RT are slow, supporting hardware RT feature make sense since nVidia has hardware RT, and soon AMD will have card with hardware RT support , same with Intel Discrete GPU. I've tried other engines with ray tracing support via software(read not using RT hardware) and they are extremely slow, like 4 frames per second compare to 20 fps on Unreal Engine 4 with DXR on a RTX 2080Ti. I don't mind if Armory engine becomes the "Crysis" of the game engine world, catering to only high end graphics. So many engines already support mobile game development. 

Another question is should Voxel Cone Tracing be dropped also? It's got too many problems and ray tracing hardware will become common place in this year.

**Comment by trsh** [comment]
> Another question is should Voxel Cone Tracing be dropped also? It's got too many problems and ray tracing hardware will become common place in this year.

Only for real game Geeks. What is like 20%? Most casual gamers (like me) will wait for years until the RTX card price drops to something reasonable. I don't suggest to make Raytracing via Vulkan orsmth, I just said it's possible. I suggest to postpone it until it's more _mainstream_. And it would be a bit stupid to drop low-end lighting, as we still want to develop for mobiles and consoles like Switch :P (and WEB).. , not that armory becomes some monster only for 3000$ pc's and next gen consoles.

**Comment by SonKim** [comment]
> > Another question is should Voxel Cone Tracing be dropped also? It's got too many problems and ray tracing hardware will become common place in this year.
> 
> Only for real game Geeks. What is like 20%? Most casual gamers (like me) will wait for years until the RTX card price drops to something reasonable. I don't suggest to make Raytracing via Vulkan orsmth, I just said it's possible. I suggest to postpone it until it's more _mainstream_. And it would be a bit stupid to drop low-end lighting, as we still want to develop for mobiles and consoles like Switch :P (and WEB).. , not that armory becomes some monster only for 3000$ pc's and next gen consoles. Graphics scalebility is every engine strength. Look at UE - one can develop games almost for all platforms with beautiful (appropriate for hardware) graphics.

You have to realize a game engine are used for many fields nowdays, look at UE4. Alot of people are using it outside the game space, like interactive real-time walk through of homes, product shots,automotive configuration,even still frame rendering,etc. There are even studios using for animated cartoons(I think Barbie uses UE4). That is why Epics Game isn't waiting for ray tracing to become cheap or mainstream they are actively working on improving ray tracing now because their customers want it and demand it.  I can see Armory being used outside the gaming space too just like UE4.

**Comment by trsh** [comment]
So bottom line, work on tracers but don't drop low-end lighting. Not up here for long discussions. I think @luboslenco already has a plan and will stick to it.

**Comment by SonKim** [comment]
> So bottom line, work on tracers but don't drop low-end lighting. Not up here for long discussions. I think @luboslenco already has a plan and will stick to it.

I could care less if low-end lighting is drop, it's not that important to me. I can definitely see its usefulness for creating games for devices like IPad's. The wording about dropping older API made it sound like lubos was dropping support for low end platforms, that's the only reason I brought it up.

**Comment by Nodragem** [comment]
I think you need to communicate more with your community, because many people do not know about your monthly releases ^^
See this recent video on Armory: https://youtu.be/lwdVwMTnCm4

**Comment by trsh** [comment]
Video on point!

**Comment by onelsonic** [comment]
As of today the gaming market:
DXR : Justified for Xbox only (has RTX implementation)
Vulkan : for mostly everything else (Win/MacOs/Linux/Android/iOS/Switch) achieving higher FPS than DXR on PC. Running on top of Metal on Apple devices. Avantage is same code used.
Gnm/Gnmx and PSSL : only PS4
Metal : only Apple devices.
WebGPU : everything with a web browser.

Hardware needed to run a realtime Raytracing demo at 24fps : a DGX station which is
x4 four Tesla V100 (+50,000$).

Raytracing has no comprehensive material library as of today with proper reflexion/refraction/scattering (Just accelerates current material rendering)

Current market data : 
mobile iOS gaming together with Android gaming now generates more revenue than Playstation+Xbox+Nintendo together.

Vulkan and WebGPU are the clear winners in terms of targeted systems and complexity meaning less bugs to fix and more systems to target.
For Raytracing focusing on deep learning noise removal by training from your own game rendering data will give you better and quicker results and it can be done using Vulkan and WebGPU.

For Xbox X/PS5 you better stick with UE4 for now before switching to the cloud.

Networking and multiplayer is also important.

**Comment by SonKim** [comment]
> Hardware needed to run a realtime Raytracing demo at 24fps : a DGX station which is
> x4 four Tesla V100 (+50,000$).


No you can running real time ray tracing on a single RTX 2080Ti now, which is cheap at 1,200 USD. RTX cards are also great for rendering offline. Performance in terms of frame rate is always  something the UE4 programmers work on with every release. With UE4.24 we got better performing real-time Gi via final gathering. I expect more performance improvement on the software side, that along with the new RTX 3000 GPU this year should make real-time ray tracing ever better. Support for Multiple GPU is also something they want to add in the future.

**Comment by trsh** [comment]
RTX 2080Ti 1500+ EUR on my country. That is not cheap for non gaming geeks. And then u need a mother board and CPU that compliments this GPU. If it's cheap for you, fine, not for me and many others. Can't Imagine kids getting this card under Christmas tree in my country. One can by a decent Car for that money.

**Comment by onelsonic** [comment]
RTX 2080Ti don't have enough RAM to run smoothly without overheating.

**Comment by SonKim** [comment]
> RTX 2080Ti don't have enough RAM to run smoothly without overheating and we are below 24fps with Raytracing ...
> You need a Quadro or SLI for 24fps, still not for everyone.
> 
> While the RTX 2080ti is the perfect raster card.

Epic Games show off the Star Wars demo with a RTX 2080 Ti it ran just fine and that was a year after they show off the demo on a DGX

**Comment by SonKim** [comment]
> RTX 2080Ti 1500+ EUR on my country. That is not cheap for casual gamers. And then u need a mother board and CPU that compliments this GPU. If it's cheap for you, fine, not for me and many others. Can't Imagine kids getting this card under Christmas tree in my country. One can buy a decent used Car for that money, lol.

I understand it's not cheap for everyone, different standard of living. It's cheap for me only because I'm a freelance artist, so I make money from my 2 RTX 2080 Ti. I use it for rendering in Blender. the RT core definitely speed up rendering. Buying the RTX 2080 Ti for gaming is is bad idea I agree.

Like i said game engine aren't for just games nowadays, its becoming increasingly popular for artist to create beautiful pixels.  Even Unity added support for ray tracing via DXR, so I would to see ray tracing support with Armory this year or next. Armory is kind of special in that its integrated into Blender so people who want to render animation/still can do it all in Blender once real-time ray tracing is implemented, no need to export out to Unity/UE4/Godot.

**Comment by onelsonic** [comment]
> > RTX 2080Ti don't have enough RAM to run smoothly without overheating and we are below 24fps with Raytracing ...
> > You need a Quadro or SLI for 24fps, still not for everyone.
> > While the RTX 2080ti is the perfect raster card.
> 
> Epic Games show off the Star Wars demo with a RTX 2080 Ti it ran just fine and that was a year after they show off the demo on a DGX

That was running on a 10,000$ Quatro RTX, otherwise please send link, I would enjoy to test UE4 with 2080Ti if we can do it.

**Comment by trsh** [comment]
Guys take this to the discord or forums pls :(

**Comment by daniel-heywood** [comment]
With Kha, there was the ability to "cross-compile your code and optimize your assets for even the most obscure systems. ..." Does this change that?

**Comment by RobDangerous** [comment]
Armory was always a little restrictive in regards to the targets supported by Kha because the physics engine was only ever supported in C/C++ and JavaScript, therefore switching from Kha to Kinc does not in itself change anything in regards to the supported targets.

**Comment by Disar** [comment]
@RobDangerous  moving forward with G5 however i guess gles2/webgl will be cut?

**Comment by RobDangerous** [comment]
Not because of G5, that also works on gles2/webgl and will continue to do so. G5 is not about new GPU features - I repeat that quite often but nobody ever listens.

**Comment by nanjizal** [comment]
I expect this could be used with current Armory it is not tied to c++ or js or even Kha, I have not tried it yet. 

https://github.com/saharan/OimoPhysics

**Comment by onelsonic** [comment]
@nanjizal there's an haxe rewrite of this AS3 flash code here :
https://github.com/armory3d/oimo_module
Going forward my guess is that we could use physic engines written in C.

**Comment by karlxuthus** [comment]
Hi，I can see Armory3d  will become the engine in other fields. I can use it to make the visualization project like more complex Verge3d(webgl). Maybe I can use it with MAX/MSP to create media interaction system. In addition, it is easy to write UI interface.

**Comment by Simonrazer** [comment]
@luboslenco Can we please have an update on the progress? Will Armory take advantage of armcore?

**Comment by luboslenco** [comment]
The original plan stands intact, I have been working on the [D3D12, Vulkan and Metal backends](https://github.com/Kode/Kinc/pulls?q=is%3Apr+author%3Aluboslenco), next up is
> Move low-level parts of Armory (like [iron](https://github.com/armory3d/iron)) to C. This is to better take advantage of multi-threading coupled with Graphics5 API. As a result, Armory will not be dependant on Kha (Haxe), but will interface directly with Kinc (C). An [extended version of Krom](https://github.com/armory3d/Kromx) which exposes Armory-specific functionality from C to Haxe/JS will be developed to accomplish this.

**Comment by MoritzBrueckner** [comment]
That's good news!

Btw, the current Kromx(armorcore) binaries for Armory are a bit out of date (VR is broken for example due to the latest changes in Iron and Kromx), what do you think of implementing automatic builds via Github actions so that the binaries are up to date for each monthly release at least?

**Comment by luboslenco** [comment]
It should be done asap once things settle down but right now it would be too much for me to keep up. I believe VR was already busted, but it will work better via armorcore/Kinc once finished.

**Comment by mundusnine** [comment]
Are there any plans on still supporting hxcpp as a target with iron in c ? The focus of armory/armorpaint seems mostly focused on Krom/armorcore and before that on HL/C. Since hxcpp's GC is better then HL/C's GC and the target that is used on consoles right now with Kha/Kinc is hxcpp maybe more emphasis should be put on that target. Also, Armorcore couldn't be utilised on consoles since the only option would be to use jitless but that would have huge performance costs.

**Comment by luboslenco** [comment]
Armory will not be dependent on Kha (Haxe) once ArmorCore is finished, hence no hxcpp / hlc (Kinc has no ties to these). It will be hooked directly to Kinc, all of the engine heavy lifting will be done in C. The option to write traits in Haxe/logic-nodes will stay, but it will also be possible to write traits in C.

> Move low-level parts of Armory (like [iron](https://github.com/armory3d/iron)) to C. This is to better take advantage of multi-threading coupled with Graphics5 API. As a result, Armory will not be dependant on Kha (Haxe), but will interface directly with Kinc (C). An [extended version of Krom](https://github.com/armory3d/Kromx) which exposes Armory-specific functionality from C to Haxe/JS will be developed to accomplish this.

**Comment by mundusnine** [comment]
I will word the question differently:

If I write haxe traits now, could I be able to use that code on pure native targets. My understanding from what you said is no because haxe will only be supported on Krom.

The other question that arises is, how will zui fit into this ? If zui can only be used on the Krom target it wont be functional for native targets for game UI's.

**Comment by luboslenco** [comment]
Can you elaborate on what do you mean by pure native targets? There will be only one 'target', hence the move to C - only the game logic will be (if you choose so) in Haxe. Just to clarify - Krom and ArmorCore will not be the same thing (that's why I renamed it from Kromx to make it clearer). ArmorCore will expose interface specifically for Armory.

**Comment by mundusnine** [comment]
`There will be only one 'target', hence the move to C - only the game logic will be (if you choose so) in Haxe.`<----But will the haxe be compiled to js (to be used with Armorcore V8) or cpp with armorcore interfaced with hxcpp GC ? Because haxe needs a GC. So the only language target for haxe would be js ? 

A pure native target would be a target that doesn't use V8 and is compiled to c/cpp(HL or hxcpp).

**Comment by luboslenco** [comment]
No hxcpp / hlc like mentioned - if one wants the GC then the game logic goes through the V8. But if you want to sidestep V8 & GC, you can - via C, Rust, or whatever compiles to WebAssembly (which can be converted to C). It's designed with the C/Kinc in mind first + the ability to sidestep GC.

**Comment by mundusnine** [comment]
Ok when you say we can use anything that compiles to WebAssembly so it can then be converted to C you mean by using this [wasm2c](https://chromium.googlesource.com/external/github.com/WebAssembly/wabt/+/HEAD/wasm2c/README.md) ?

On another subject, if it's the case to prioritize native is their any plans to have access to armorcore with library compilation(by using KINC_FUNC before native functions) so we can use armory from native directly ? Like what ustance and me are doing with our wrapping of kinc to other [native langs](https://github.com/Kode-Community). 

Will the webassembly 2 c be preferred as opposed to giving access from a lib to native langs ?

**Comment by N8n5h** [comment]
What it's the plan for console support? @luboslenco

**Comment by luboslenco** [comment]
> Ok when you say we can use anything that compiles to WebAssembly so it can then be converted to C you mean by using this wasm2c ?

A tool like that or [similar](https://github.com/appcypher/awesome-wasm-runtimes), might be also possible to "exploit" the v8 turbofan for AOT compilation, maybe wasm-to-llvm option eventually pops up. 

> is their any plans to have access to armorcore/armory with library compilation

Right now there is plenty on the plate but yes would be nice to do that eventually.

> It just seems unclear what the road to this is considering armory is right now pretty much Haxe bound

The armory repository should stay intact for the most part, I will take care of the transition where necessary (ie. rewire calls directly to armorcore once ready).

> What it's the plan for console support?

The plan is to use Kinc as usual for that. As for writing logic - it will be ok to use logic nodes, but if the game needs complex logic it will be recommended to use non-jit / non-gc language for heavy stuff.

**Comment by kariem2k** [comment]
@luboslenco Thanks for the clarification, but I still don't know how traits can be written in C/C++.  The C trait API is very basic https://github.com/armory3d/armory/blob/master/Sources/armory/trait/internal/wasm_api.h ? Am I missing something? or the API still needs to be added here? 

Thanks

**Comment by luboslenco** [comment]
Sorry the API is not yet available. First the iron library has to be ported, working on it.

**Comment by kariem2k** [comment]
Thank you very much for your reply. I was just asking if I am missing something. Thank you for all of this great and huge undertaking.

**Comment by Joe23232** [comment]
I created an issue https://github.com/armory3d/armory/issues/2016#issue-745401970 where I believe the Rust code for web assembly should be more simplified. There is just so much text going on.

**Comment by onelsonic** [comment]
@Joe23232 the WASM code looks already pretty neat compare to pure rust and is working.
https://github.com/gwowen/rust-3D-cube

**Comment by Joe23232** [comment]
@onelsonic Is this your code?

**Comment by trsh** [comment]
@luboslenco 2020 is almost over. Can you inform the community of what has been done and what needs to be done in terms of these plans? I think we deserve some honesty.

**Comment by Lykdraft** [comment]
> @luboslenco 2020 is almost over. Can you inform the community of what has been done and what needs to be done in terms of these plans? I think we deserve some honesty.

This!

**Comment by mundusnine** [comment]
> 
> 
> Sorry the API is not yet available. First the iron library has to be ported, working on it.

@luboslenco 
Anything public regarding this ? Started doing it on my side at the moment: https://github.com/Kode-Community/ironc

If ever you have already something I would just add to it.
